### PR TITLE
Prevent overwrites on merge

### PR DESCRIPTION
**Pull request by jacoblurye**
This PR introduces the `ThrowOnCollision` merge strategy as the default merge strategy for literal values in `merge_clinical_trial_metadata`. This strategy throws an error on attempted updates to properties on existing trials/participants/samples/etc. - some existing tests were updated to accommodate this change.

### CODE REVIEW

File: cidc_schemas/prism.py [hunk 1]
```diff
@@ -808,6 +808,38 @@ def _update_artifact(ct: dict, artifact_patch: dict, artifact_uuid: str) -> (dic
...
+    Similar to the jsonmerge's built in 'discard' strategy,
+    but throws an error if the value already exists. This is
+    used to prevent updates in `merge_clinical_trial_metadata`.
+    """
+
+    def merge(self, walk, base, head, schema, meta, **kwargs):
+        if base.is_undef():
+            return head
+        if base.val != head.val:
+            prop = base.ref.split("/")[-1]
```
Comment (curlup) [inline cidc_schemas/prism.py]: nit:
```suggestion
            prop = base.ref.rsplit("/",1)[-1]
```
due to 
```
In [1]: %timeit "/whatever/this/is/1".split("/")[-1]
209 ns ± 1.91 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [2]: %timeit "/whatever/this/is/1".rsplit("/",1)[-1]
172 ns ± 3.96 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

File: tests/test_prism.py [hunk 2]
```diff
@@ -21,7 +21,8 @@
 from cidc_schemas.prism import prismify, merge_artifact, \
     merge_clinical_trial_metadata, InvalidMergeTargetException, \
     SUPPORTED_ASSAYS, SUPPORTED_MANIFESTS, SUPPORTED_TEMPLATES, \
-    PROTOCOL_ID_FIELD_NAME, parse_npx, merge_artifact_extra_metadata
+    PROTOCOL_ID_FIELD_NAME, parse_npx, merge_artifact_extra_metadata, \
+     PRISM_STRATEGIES, ThrowOnCollision, MergeCollisionException
```
Comment (curlup) [inline tests/test_prism.py]: nit: indent seems strange

File: tests/test_prism.py [hunk 3]
```diff
@@ -1102,3 +1100,43 @@ def test_parse_npx_merged(npx_combined_file_path):
...
+
+    merger = Merger(schema, strategies=PRISM_STRATEGIES)
+
+    # Identical values, no collision - no error
+    base = {'l': [{'cimac_id': 'c1', 'a': 1}]}
+    assert merger.merge(base, base)
+
+    # Different values, collision - error
+    head = {'l': [{'cimac_id': 'c1', 'a': 2}]}
+    with pytest.raises(MergeCollisionException):
```
Comment (curlup) [inline tests/test_prism.py]: consider adding `match` param to check that it reports the collision you expect.

File: cidc_schemas/prism.py [hunk 4]
```diff
@@ -808,6 +808,38 @@ def _update_artifact(ct: dict, artifact_patch: dict, artifact_uuid: str) -> (dic
...
+    """
+    Similar to the jsonmerge's built in 'discard' strategy,
+    but throws an error if the value already exists. This is
+    used to prevent updates in `merge_clinical_trial_metadata`.
+    """
+
+    def merge(self, walk, base, head, schema, meta, **kwargs):
+        if base.is_undef():
+            return head
+        if base.val != head.val:
```
Comment (curlup) [inline cidc_schemas/prism.py]: just to clear it up - you don't check for `head.is_undef` here, because it won't get called on that?

Comment (jacoblurye) [reply cidc_schemas/prism.py]: That's my understanding, yes.

**Pull request merged**